### DISCUSSION TITLE
How to get at TreeViewItem from within TreeView.ItemTemplate?

### DISCUSSION THREAD

**Discussion by fitdev** [AvaloniaUI/Avalonia#11845] [Q&A] [answered]
While I can easily construct the _interior_ of `TreeViewItem`s using `<TreeView.ItemTemplate>`, how can I affect the properties of the containing `TreeViewItem`s themselves, especially with the ability to bind some of its properties to those coming from `ItemTemplate`' associated type?

And more generally, this question applies to any `ItemsCOntrol` not just `TreeView`.

**Comment by timunie** [comment]
the question is not clear enough. Why exactly do you need to access the control? In common cases, the ItemViewModel can have a reference to the MainViewModel and thus it's knowing the parent. Or you can bind to parent control or named control if really needed. Please discribe the use case and the expected outcome more clearly. 

Thank you.

**Reply by fitdev** [reply]
For example:

```xml
<TreeView ItemsSource="{Binding TreeRootItems}">
	  <TreeView.ItemTemplate>
                  <!-- How can I affect the TreeViewItem that will receive the contents below from within the template? I mean what if I want to change some of its properties like Visibility, Height, etc. - something I cannot do by just changing its contents? -->
		  <TreeDataTemplate DataType="mod:TreeNodeModel" ItemsSource="{Binding Subitems}">
			  <TextBlock Text="{Binding MyNodeText}"/>
		  </TreeDataTemplate>
	  </TreeView.ItemTemplate>
</TreeView>

```

**Reply by rabbitism** [reply]
easiest way is to define ItemContainerTheme and bind to DataContext.YourProperty.

**Reply by fitdev** [reply]
@rabbitism Can you please provide a quick snippet as to how to do it?

**Reply by timunie** [reply]
Option 1: `{Binding $parent[TreeView].((vm:MyViewModel)DataContext).MyProperty}`
Option 2: `{Binding #TheNameOfMyRootControl.((vm:MyViewModel)DataContext).MyProperty}`

**Reply by fitdev** [reply]
@timunie Thank you once again for your help. But where would I insert either of this in my snippet? In the `TreeDataTemplate` element? And which attribute do I set then?

For example, suppose I want to set the `Cursor` property of the `TreeViewItem` for whose `TreeView` I have defined the above `TreeDataTemplate` (this is just an example, I know I can set `Cursor` on the controls inside of my data template, but suppose I need to do it on the parent `TreeViewItem` instead). How do I do it then?

**Reply by timunie** [reply]
Can also be in the container or wherever you need it. Up to you.

**Comment by fitdev** [comment]
I am still at a loss. Surely it is not like this (assuming the `Cursor` property of `TreeViewItem` I have above):

```xml
<TreeView ItemsSource="{Binding TreeRootItems}">
	  <TreeView.ItemTemplate>
		  <TreeDataTemplate DataType="mod:TreeNodeModel" ItemsSource="{Binding Subitems}" TreeViewItem.Cursor="{Binding $parent[TreeViewItem].((vm:MyViewModel)DataContext).MyCursorProperty}">
			  <TextBlock Text="{Binding MyNodeText}"/>
		  </TreeDataTemplate>
	  </TreeView.ItemTemplate>
</TreeView>
```

**Reply by timunie** [reply]
Can you please file a minimal sample of what you have and give me a hint what is the goal you have? We are not making any progress atm as I don't know what exactly to do.

**Reply by fitdev** [reply]
Here's the repro. I have added a few comments prefixed with `QUESTION` in `MainWindow.axaml` and `MainWindow.axaml.cs` files.

[AvaloniaItemAccessFromTemplateSample.zip](https://github.com/AvaloniaUI/Avalonia/files/11816030/AvaloniaItemAccessFromTemplateSample.zip)

**Reply by timunie** [reply]
that was useful 👍

**Comment by timunie** [comment, marked as answer]
looking at the sample you need to do two things

## 1 Use DataTemplates instead of filing controls for sub items:

```cs
public IDataTemplate? CategoryContentTemplate { get; set; }

public IDataTemplate? ItemContentTemplate { get; set; }

public Control? Build(object? data)
{

    var visualTreeForItem = (CategoryDataType?.IsInstanceOfType(data) ?? false)
            ? CategoryContentTemplate?.Build(data)
            : ((ItemDataType?.IsInstanceOfType(data) ?? false) ? ItemContentTemplate?.Build(data) : null);

    return visualTreeForItem;
}
```

and in xaml:

```xml
<TreeView.ItemTemplate>
	<vm:MyTreeDataTemplate
		CategoryDataType="vm:TreeNodeModel"
		ItemDataType="vm:TreeSubNodeModel"
		x:DataType="vm:TreeNodeModel"
		ItemsSource="{Binding Subitems}">
		<vm:MyTreeDataTemplate.CategoryContentTemplate>
			<DataTemplate DataType="vm:TreeNodeModel">
				<TextBlock Text="{Binding Name}"
                                           Background="Red"/>
			</DataTemplate>
		</vm:MyTreeDataTemplate.CategoryContentTemplate>
		<vm:MyTreeDataTemplate.ItemContentTemplate>
			<DataTemplate DataType="vm:TreeSubNodeModel">
				<TextBlock Height="{Binding Height}" 
                                           Text="{Binding Name}" 
                                           IsVisible="{Binding Visible}" />
			</DataTemplate>
		</vm:MyTreeDataTemplate.ItemContentTemplate>
	</vm:MyTreeDataTemplate>
</TreeView.ItemTemplate>
```

## 2 For accessing ItemContainer create a ControlTheme

```xml
<TreeView.ItemContainerTheme>
	<ControlTheme TargetType="TreeViewItem" BasedOn="{StaticResource {x:Type TreeViewItem}}">
		<Setter Property="IsVisible" Value="{ReflectionBinding Visible}" />
	</ControlTheme>
</TreeView.ItemContainerTheme>
```

Note: I'm using `ReflectionBinding` as I don't know the datatype in advance. You can use `CompiledBindings` if you use create a converter or extract a common base class or common interface.

**Reply by fitdev** [reply]
Thanks so much for your help @timunie! Works great!

Just a few questions:

1. Your suggestion regarding using `IDataTemplate` - is it merely a matter of preference, or is it because that's what Avalonia expects?
2. Incidentally is there a way to avoid using another private instance of `TreeDataTemplate` to which we delegate the implementation of `InstancedBinding? ITreeDataTemplate.ItemsSelector(object item)`? Perhaps `TreeDataTemplate`'s implementation of it can be made available to the outside callers as a static method: `public static InstancedBinding? ItemsSelector(BindingBase? itemSource, object item)`?
3. I placed `<TreeView.ItemContainerTheme>` right within `<TreeView>` but I am not sure how it works. Does it then only affect a concrete `TreeView`?
4. I could not find `ItemContainerTheme` in the docs anywhere, how exactly does it work, and couldn't `Styles` somehow be used for that instead? How is the theme different from styles then?

Sorry for all the questions. I am just beginning to learn Avalonia.

**Reply by timunie** [reply]
1 -> Better to use Templates (DataTemplates, ControlTemplates etc). You ask for pain if not (having two tabs would break the app for example)

2 -> i think you don't need it at all with proper design. 

3 -> in this case, yes. You can also add it to a Style and then you have it everywhere

4 -> yeah, you use a preview version. Docs are for stable version ... We are working on it rn.
